pkg/cluster/base.go

```diff
@@ -364,6 +374,86 @@ func (r *CreateClusterRequest) validateMainFields() error {
 	return nil
 }
 
+// constants for desiredSpotPrice validation
+const (
+	maxZonePrice     = "maxZonePrice"
+	desiredSpotPrice = "desiredSpotPrice"
+)
+
+func (r *CreateClusterRequest) validateAmazonSpotPrices() error {
+	var errs []error
+	if r.Properties.CreateClusterPKE != nil {
+		for _, np := range r.Properties.CreateClusterPKE.NodePools {
+			providerConfig := internalPke.NodePoolProviderConfigAmazon{}
+			if err := mapstructure.Decode(np.ProviderConfig, &providerConfig); err != nil {
+				return errors.WrapIff(err, "decoding nodepool %q config", np.Name)
+			}
+
+			// check if spotPrice is specified
+			if providerConfig.AutoScalingGroup.SpotPrice == "" {
+				continue
+			}
+
+			spotPriceMap, err := r.getSpotPriceMap(providerConfig.AutoScalingGroup.SpotPrice, cloud.Amazon, PKE, providerConfig.AutoScalingGroup.InstanceType)
+			if err != nil {
+				return err
+			}
+
+			for _, zone := range providerConfig.AutoScalingGroup.Zones {
+				if spotPriceMap[desiredSpotPrice] < spotPriceMap[string(zone)] {
+					errs = append(errs, pkgErrors.NewSpotPriceError(spotPriceMap[desiredSpotPrice], spotPriceMap[string(zone)], string(zone), providerConfig.AutoScalingGroup.InstanceType))
+				}
+			}
+
+			if errs == nil && spotPriceMap[desiredSpotPrice] < spotPriceMap[maxZonePrice] {
+				errs = append(errs, pkgErrors.NewSpotPriceError(spotPriceMap[desiredSpotPrice], spotPriceMap[maxZonePrice], r.Location, providerConfig.AutoScalingGroup.InstanceType))
+			}
+		}
+	} else {
+		for _, np := range r.Properties.CreateClusterEKS.NodePools {
+			// check if spotPrice is specified
+			if np.SpotPrice == "" {
+				continue
+			}
+
+			spotPriceMap, err := r.getSpotPriceMap(np.SpotPrice, cloud.Amazon, EKS, np.InstanceType)
+			if err != nil {
+				return err
+			}
+
+			if spotPriceMap[desiredSpotPrice] < spotPriceMap[maxZonePrice] {
+				errs = append(errs, pkgErrors.NewSpotPriceError(spotPriceMap[desiredSpotPrice], spotPriceMap[maxZonePrice], r.Location, np.InstanceType))
+			}
+		}
+	}
```

Comment:
**Required**: as I mentioned in private, this very big 2 if branch can be simplified into a single for loop, let's discuss it offline.

Reply:
Discussed, for the time being there is no easier alternative because of a bunch of different types.